crates/sui-network/src/state_sync/mod.rs

```diff
@@ -1146,32 +1147,43 @@ async fn sync_checkpoint_contents_from_archive<S>(
         };
         debug!("Syncing checkpoint contents from archive: {sync_from_archive},  highest_synced: {highest_synced},  lowest_checkpoint_on_peers: {}", lowest_checkpoint_on_peers.map_or_else(|| "None".to_string(), |l| l.to_string()));
         if sync_from_archive {
-            let start = highest_synced
-                .checked_add(1)
-                .expect("Checkpoint seq num overflow");
-            let checkpoint_range = start..lowest_checkpoint_on_peers.unwrap();
-            if let Some(archive_reader) = archive_readers
-                .pick_one_random(checkpoint_range.clone())
+            'inner: {
+                let start = highest_synced
+                    .checked_add(1)
+                    .expect("Checkpoint seq num overflow");
+                let end = lowest_checkpoint_on_peers.unwrap();
+
+                let Some(ref archive_config) = archive_config else {
+                    warn!("Failed to find an archive reader to complete the state sync request");
+                    break 'inner;
+                };
+                let Some(ref ingestion_url) = archive_config.ingestion_url else {
+                    warn!("Archival ingestion url for state sync is not configured");
+                    break 'inner;
+                };
+                let reader_options = ReaderOptions {
+                    batch_size: archive_config.download_concurrency.into(),
+                    upper_limit: Some(end),
+                    ..Default::default()
+                };
+                let Ok((executor, _exit_sender)) = setup_single_workflow(
+                    StateSyncWorker(store.clone()),
+                    ingestion_url.clone(),
+                    start,
+                    1,
+                    Some(reader_options),
+                )
                 .await
-            {
-                let txn_counter = Arc::new(AtomicU64::new(0));
-                let checkpoint_counter = Arc::new(AtomicU64::new(0));
-                if let Err(err) = archive_reader
-                    .read(
-                        store.clone(),
-                        checkpoint_range,
-                        txn_counter.clone(),
-                        checkpoint_counter.clone(),
-                        true,
-                    )
-                    .await
-                {
-                    warn!("State sync from archive failed with error: {:?}", err);
-                } else {
-                    info!("State sync from archive is complete. Checkpoints downloaded = {:?}, Txns downloaded = {:?}", checkpoint_counter.load(Ordering::Relaxed), txn_counter.load(Ordering::Relaxed));
+                else {
+                    break 'inner;
+                };
```

Comment:
I'm having a little bit of trouble understanding the `break 'inner` control flow here. would this maybe be clearer if it was a separate function that could early return or something?

Reply:
it just skips to the last statement in the loop(i.e. sleep), but I don't mind